homeassistant/components/hydrawise.py

```diff
@@ -0,0 +1,154 @@
+    'next_cycle': ['Next Cycle', 'mdi:calendar-clock', '', ''],
+    'status': ['Status', '', 'connectivity', ''],
+    'watering_time': ['Watering Time', 'mdi:water-pump', '', 'min'],
+    'rain_sensor': ['Rain Sensor', '', 'moisture', '']
+}
+
+BINARY_SENSORS = ['is_watering', 'status', 'rain_sensor']
+
+SENSORS = ['next_cycle', 'watering_time']
+
+SWITCHES = ['auto_watering', 'manual_watering']
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+SIGNAL_UPDATE_HYDRAWISE = "hydrawise_update"
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_ACCESS_TOKEN): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL, default=SCAN_INTERVAL):
+            cv.time_period,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the Hunter Hydrawise component."""
+    conf = config[DOMAIN]
+    access_token = conf.get(CONF_ACCESS_TOKEN)
+    scan_interval = conf.get(CONF_SCAN_INTERVAL)
+
+    try:
+        from hydrawiser.core import Hydrawiser
+
+        hydrawise = Hydrawiser(user_token=access_token)
+        hass.data[DATA_HYDRAWISE] = HydrawiseHub(hydrawise)
+    except (ConnectTimeout, HTTPError) as ex:
+        _LOGGER.error(
+            "Unable to connect to Hydrawise cloud service: %s", str(ex))
+        hass.components.persistent_notification.create(
+            'Error: {}<br />'
+            'You will need to restart hass after fixing.'
+            ''.format(ex),
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID)
+        return False
+
+    def hub_refresh(event_time):
+        """Call Hydrawise hub to refresh information."""
+        _LOGGER.debug("Updating Hydrawise Hub component")
+        hass.data[DATA_HYDRAWISE].data.update_controller_info()
+        dispatcher_send(hass, SIGNAL_UPDATE_HYDRAWISE)
+
+    # Call the Hydrawise API to refresh updates
+    track_time_interval(hass, hub_refresh, scan_interval)
+
+    return True
+
+
+class HydrawiseHub(object):
+    """Representation of a base Hydrawise device."""
+
+    def __init__(self, data):
+        """Initialize the entity."""
+        self.data = data
+
+
+class HydrawiseEntity(Entity):
+    """Entity class for Hydrawise devices."""
+
+    def __init__(self, data, sensor_type):
+        """Initialize the Hydrawise entity."""
+        self.data = data
+        self._sensor_type = sensor_type
+        self._name = "{0} {1}".format(
+            self.data.get('name'),
+            DEVICE_MAP.get(self._sensor_type)[
+                DEVICE_MAP_INDEX.index('KEY_INDEX')])
+        self._state = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        async_dispatcher_connect(
+            self.hass, SIGNAL_UPDATE_HYDRAWISE, self._update_callback)
+
+    def _update_callback(self):
```

Comment:
Decorate this with `@callback` imported from core.py.

Reply:
Done.